packages/core/src/modules/drpc/DRPCMessagesApi.ts

```diff
@@ -0,0 +1,173 @@
+
+@injectable()
+export class DRPCMessagesApi {
+  private drpcMessageService: DRPCMessageService
+  private drpcMethodHandlers: Map<
+    string,
+    (message: DRPCRequestObject) => Promise<DRPCResponseObject | Record<string, never>>
+  > = new Map()
+  private messageSender: MessageSender
+  private connectionService: ConnectionService
+  private agentContext: AgentContext
+
+  public constructor(
+    messageHandlerRegistry: MessageHandlerRegistry,
+    drpcMessageService: DRPCMessageService,
+    messageSender: MessageSender,
+    connectionService: ConnectionService,
+    agentContext: AgentContext
+  ) {
+    this.drpcMessageService = drpcMessageService
+    this.messageSender = messageSender
+    this.connectionService = connectionService
+    this.agentContext = agentContext
+    this.registerMessageHandlers(messageHandlerRegistry)
+
+    // listen for incoming drpc requests and forward them to appropriate handlers
+    this.drpcMessageService.createMessageListener(async ({ message, drpcMessageRecord }) => {
+      if (message instanceof DRPCRequestMessage && drpcMessageRecord.role === DRPCMessageRole.Receiver) {
+        const requests = Array.isArray(message.request) ? message.request : [message.request]
+        const futures = []
+        for (const request of requests) {
+          futures.push(this.handleRequest(request))
+        }
+        const responses = await Promise.all(futures)
+        try {
+          await this.sendDRPCResponse(
+            drpcMessageRecord.connectionId,
+            message.id,
+            responses.length === 1 ? responses[0] : responses
+          )
+        } catch {
+          await this.sendDRPCResponse(drpcMessageRecord.connectionId, message.id, {
+            jsonrpc: '2.0',
+            id: null,
+            error: { code: DRPCErrorCode.INTERNAL_ERROR, message: 'Internal error', data: 'Error sending response' },
+          })
+        }
+      }
+    })
+  }
+
+  private async handleRequest(message: DRPCRequestObject) {
+    if (!isValidDRPCRequestObject(message)) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.INVALID_REQUEST, message: 'Invalid request' },
+      }
+    }
+    const handler = this.drpcMethodHandlers.get(message.method)
+    if (!handler) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.METHOD_NOT_FOUND, message: 'Method not found' },
+      }
+    }
+
+    try {
+      let response: DRPCResponseObject | Record<string, never> = await handler(message)
+      if (message.id === null) {
+        response = {}
+      }
+      return response
+    } catch (error) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.INTERNAL_ERROR, message: 'Internal error', data: error.message },
+      }
+    }
+  }
+
+  public async sendDRPCRequest(connectionId: string, request: DRPCRequest): Promise<DRPCResponse> {
```

Comment:
Since you are awaiting for a response, IMHO there should be a timeout setting. Otherwise, the execution will block in case the other party does not respond. Maybe it can follow the pattern we do in DiscoverFeaturesApi's `queryFeatures()` ( `awaitDisclosures` setting).

Reply:
Is rpc meant to be request -> response, or also to support async flows? If request/ response i think timeout should suffice (which we can make configurable on the module and request method). Otherwise async (response should be catched using events) may work better.

Reply:
The way it is implemented in ACA-Py and described in the RFC (at least how I've interpreted) it is async.